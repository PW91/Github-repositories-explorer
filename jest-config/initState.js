export default {
  ui: {
    activeUserCard: null,
    lastSearchQuery: null,
    isSearchEmpty: null,
    isLoading: null
  },
  users: []
}