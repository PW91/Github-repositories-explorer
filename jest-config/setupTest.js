import Enzyme from 'enzyme'
import EnzymeAdapter from 'enzyme-adapter-react-16'
import 'regenerator-runtime/runtime.js'
import '~/i18n'

Enzyme.configure({
  adapter: new EnzymeAdapter(),
  disableLifecycleMethods: true
})
