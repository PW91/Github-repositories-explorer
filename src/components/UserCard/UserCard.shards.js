import styled from '@emotion/styled'
import { IcoArrow } from '~/styles/icofont'
import { NoResultsContainer } from '~/components/NoResults/NoResults.shards'

export const UserCardContainer = styled.div`
  margin-bottom: 1rem;

  ${NoResultsContainer} {
    margin: .5rem 0 .5rem 2rem;
  }
`

export const UserCard = styled.div`
  padding: 1rem 2rem;
  display: flex;
  justify-content: space-between;
  align-items: center;
  cursor: pointer;
  background-color: ${({ theme }) => theme.colors.background.white};
  border-radius: ${({ theme }) => theme.borderRadius.main};
  box-shadow: ${({ theme }) => theme.shadow.box};
  border: ${({ isActive, theme }) => isActive && theme.border.element};
  transition: background-color .1s;

  &:hover {
    background-color: ${({ theme }) => theme.colors.background.lightGrey};
  }
`

export const Avatar = styled.img`
  width: 6rem;
  margin-right: 1rem;
`

export const UserMeta = styled.div`
  display: flex;
  align-items: center;
`

export const DropdownIcon = styled(IcoArrow)`
  transform: ${({ isActive }) => isActive && 'rotate(180deg)'};
  color: ${({ theme }) => theme.colors.text.main};
  font-size: ${({ theme }) => theme.fontSize.icon.normal};
  transition: transform .1s;
`
