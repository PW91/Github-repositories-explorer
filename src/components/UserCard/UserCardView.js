import React from 'react'
import PropTypes from 'prop-types'
import { useTranslation } from 'react-i18next'
import RepoCard from '~/components/RepoCard'
import NoResults from '~/components/NoResults'
import { BigHeading } from '~/components/Typography'
import {
  UserCardContainer,
  UserCard,
  UserMeta,
  DropdownIcon,
  Avatar
} from '~/components/UserCard/UserCard.shards'

const UserCardView = ({
  login,
  avatarUrl,
  isActive,
  repos,
  userCardClickFn
}) => {
  const { t } = useTranslation('common')
  return (
    <UserCardContainer data-test='UserCardView'>
      <UserCard
        onClick={userCardClickFn}
        isActive={isActive}
      >
        <UserMeta>
          <Avatar src={avatarUrl} alt='avatar' />
          <BigHeading>{login}</BigHeading>
        </UserMeta>
        <DropdownIcon isActive={isActive} />
      </UserCard>
      {isActive &&
        (repos
          ? repos.map(repo =>
            <RepoCard
              key={repo.id}
              {...repo}
            />)
          : <NoResults text={t('errors.emptyRepos')} />
        )
      }
    </UserCardContainer>
  )
}

UserCardView.propTypes = {
  login: PropTypes.string,
  avatarUrl: PropTypes.string,
  isActive: PropTypes.bool,
  repos: PropTypes.array,
  userCardClickFn: PropTypes.func
}

export default UserCardView
