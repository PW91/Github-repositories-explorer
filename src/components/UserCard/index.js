export { default } from '~/components/UserCard/UserCardView'
