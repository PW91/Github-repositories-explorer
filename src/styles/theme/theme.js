import * as colors from '~/styles/theme/colors'
import * as commons from '~/styles/theme/commons'

export default {
  colors,
  ...commons
}
