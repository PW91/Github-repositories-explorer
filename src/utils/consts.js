export const API_URL = 'https://api.github.com'

export const BREAKPOINTS = {
  SM: 480,
  MD: 768,
  LG: 992,
  XL: 1300,
}

export const SEARCH_RESULTS_LIMIT = 5

export const API_ACTIONS = {
  USERS_SEARCH: 'USERS_SEARCH',
  USER_REPOS: 'USER_REPOS'
}

export const STORE_ACTIONS = {
  SET_ACTIVE_USER_CARD: 'SET_ACTIVE_USER_CARD',
  SET_LAST_SEARCH_QUERY: 'SET_LAST_SEARCH_QUERY',
  SET_SEARCH_EMPTY_FLAG: 'SET_SEARCH_EMPTY_FLAG',
  SET_LOADING_FLAG: 'SET_LOADING_FLAG',
  SET_USERS: 'SET_USERS',
  SET_USER_REPOS: 'SET_USER_REPOS'
}
