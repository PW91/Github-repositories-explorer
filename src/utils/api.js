import { format } from 'url'
import { toast } from 'react-toastify'
import { parseJSON } from '~/utils/helpers'
import { API_URL, API_ACTIONS } from '~/utils/consts'

const ajaxer = (apiAction, path, method = 'GET', data = null, query = null) => (
  new Promise((resolve, reject) => {
    const sendForm = data && data.constructor === FormData
    const requestOptions = {
      headers: !sendForm && { 'content-type': 'application/json' },
      method,
      ...!!data && { body: sendForm ? data : JSON.stringify(data) },
    }

    const fullPath = `${API_URL}${path}${query ? format({ query }) : ''}`

    fetch(fullPath, requestOptions)
      .then(res => res.status < 400 ? res.text() : { errorText: res.statusText, status: res.status })
      .then(res => {
        if (res.errorText) {
          const { errorText, status } = res
          toast.error(errorText)
          reject({ errorText, status })
        } else {
          const parsedJson = parseJSON(res)
          const mapped = responseFormatter(parsedJson, apiAction)
          resolve(mapped)
        }
      })
  })
)

export const responseFormatter = (data, apiAction) => {
  switch (apiAction) {
    case API_ACTIONS.USERS_SEARCH:
      return data.items.map(({
        login,
        id,
        avatar_url
      }) => ({
        login,
        id,
        avatarUrl: avatar_url
      }))
    case API_ACTIONS.USER_REPOS:
      return data.map(({
        id,
        owner,
        name,
        html_url,
        stargazers_count,
        description
      }) => ({
        id,
        owner: owner.id,
        name,
        htmlUrl: html_url,
        stars: stargazers_count,
        description
      }))
  }
}

export const manageUsersSearch = {
  get: (query) => ajaxer(API_ACTIONS.USERS_SEARCH, '/search/users', 'GET', null, query)
}

export const manageUserRepos = {
  get: (username) => ajaxer(API_ACTIONS.USER_REPOS, `/users/${username}/repos`)
}
