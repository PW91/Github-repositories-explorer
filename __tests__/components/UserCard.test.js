import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import serializer from 'jest-emotion'
import { ThemeProvider } from 'emotion-theming'
import { findByTestAtrr } from '>/jest-config/utils'
import theme from '~/styles/theme/theme'
import UserCard from '~/components/UserCard'

expect.addSnapshotSerializer(serializer)

describe('UserCard component', () => {

  it('It renders', () => {
    const wrapper =  shallow(<UserCard />)
    expect(findByTestAtrr(wrapper, 'UserCardView').exists()).toBeTruthy()
  })

  it('Snapshot', () => {
    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <UserCard />
      </ThemeProvider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})