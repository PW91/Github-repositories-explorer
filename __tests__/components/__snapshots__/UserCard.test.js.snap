// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard component Snapshot 1`] = `
.emotion-10 {
  margin-bottom: 1rem;
}

.emotion-10 .e6m5mtg0 {
  margin: .5rem 0 .5rem 2rem;
}

.emotion-8 {
  padding: 1rem 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  background-color: #ffffff;
  border-radius: .2rem;
  box-shadow: 0 2px 6px 0 rgba(0,0,0,0.12);
  -webkit-transition: background-color .1s;
  transition: background-color .1s;
}

.emotion-8:hover {
  background-color: #f5f5f5;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  width: 6rem;
  margin-right: 1rem;
}

.emotion-2 {
  color: #002442;
  font-size: 2rem;
  margin: 0;
  font-weight: 700;
}

.emotion-6 {
  font-family: 'icofont' !important;
  speak: none;
  font-style: normal;
  font-weight: 400;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #002442;
  font-size: 1.8rem;
  -webkit-transition: -webkit-transform .1s;
  -webkit-transition: transform .1s;
  transition: transform .1s;
}

.emotion-6:before {
  content: "\\f101";
}

<div
  className="emotion-10 emotion-11"
  data-test="UserCardView"
>
  <div
    className="emotion-8 emotion-9"
  >
    <div
      className="emotion-4 emotion-5"
    >
      <img
        alt="avatar"
        className="emotion-0 emotion-1"
      />
      <h1
        className="emotion-2 emotion-3"
      />
    </div>
    <span
      className="emotion-6 emotion-7"
    />
  </div>
</div>
`;
